### PR TITLE
Add initial documentation

### PR DESCRIPTION
**Pull request by JamesWrigley**
Changes:
- New docs use sphinx
- Tweaked some of the existing docstrings to use LaTeX (rendered with MathJax), removed `ArrayLike` type annotations because they render absurdly long, and replaced `real` with `float` in docstrings to get some nice linking to the Python docs
- Updated the README

### CODE REVIEW

**Comment by JamesWrigley** [comment]
(if there's no objections I'll probably merge this next week)

**Comment by JamesWrigley** [comment]
In a nutshell, I didn't like it at all :man_shrugging: You can have a look at my tests in this branch: https://github.com/European-XFEL/EXtra/tree/mkdocs

I initially found it a lot more difficult to set up than sphinx because there's some weirdness around documenting module members. I tried to do this:
```
::: extra.utils
    options:
      members:
        - find_nearest_index
        - find_nearest_value
```

But it didn't display any documentation for those functions. I couldn't figure out exactly why but I suspect it's because they're actually under `extra.utils.functions` (even though they're imported in `extra.utils.__init___`). Eventually I gave up and listed them explicitly instead:
```
::: extra.utils.gaussian

::: extra.utils.gaussian2d

::: extra.utils.lorentzian
```

Then I tried documenting the `Scantool` class, and the attributes turned out like this:
image

It's putting the actual assignment statement of the attribute in the signature, which is kinda cool, but IMO it makes the signature quite unreadable. I googled for a while but haven't found a way to disable it... You can get around it by documenting the attributes in an `Attributes` section in the class docstring, but then you'd have to repeat the name and type hint.

I also prefer sphinx/RTDs theming in general, but that's probably fixable with some custom CSS rules.

Given all that, my vote would be to stick with Sphinx.

**Comment by philsmt** [comment]
Oh sorry, but thanks for trying 😢 

@RobertRosca or @D4vidH4mm3r in case they have a quick idea what could've gone wrong here, but else it's Sphinx I guess.

**Comment by JamesWrigley** [comment]
(I'll still try to merge this by the end of the week if possible :crossed_fingers: )

**Comment by JamesWrigley** [comment]
(if there's no objections I'd like to merge this tomorrow)

**Comment by RobertRosca** [comment]
Philipp just pinged me so I'll have a look at this now :+1:

**Comment by RobertRosca** [comment]
Alright I quickly whacked together a site with mkdocs to check out what's going on here https://clever-granita-6bc357.netlify.app

Few notes:

> I initially found it a lot more difficult to set up than sphinx because there's some weirdness around documenting module members. I tried to do this:

Funnily enough I never actually wrote... any code reference documentation, since I always do it automatically and prefer to keep all of the relevant module-level documentation in the python file itself instead of in a separate documentation file :see_no_evil:

Either way, my guess would be that it doesn't show up due to the start imports and the modules not being listed in `__all__`.

> Then I tried documenting the Scantool class, and the attributes turned out like this:

One thing to note is that the docstrings in this project are numpy style, but the default style for mkdocstrings is google, which is probably the main source of it looking weird. Changing that and settings a few extra defaults makes it look a[ bit better](https://clever-granita-6bc357.netlify.app/code-reference/extra/components/scantool/):

image

With the part you mentioned:

image

I didn't change any of the formatting in the docstrings to markdown/mkdocs style so references/links will be broken, but this is a rough view of what it could look like if you're interested.

**Comment by JamesWrigley** [comment]
Good catch with `__all__` :thinking: That looks way better, but I still don't think the assignment statements should be in the docs. I also kinda prefer Sphinx's theme for docs, so my vote is still for Sphinx.

**Comment by JamesWrigley** [comment]
For comparison because I finally got RTD working :sweat_smile: https://extra--1.org.readthedocs.build/en/1/index.html

**Comment by philsmt** [comment]
I still have a strong dislike for RST, but if you lean the other way, I don't feel like it's my place to decide that with all the preparations you already did. FWIW it'll look in line with EXtra-data and -geom :shrug:

**Comment by RobertRosca** [comment]
For this I don't think the choice matters too much, IMO Mkdocs makes sense if you write a lot of text (like the user documentation page) since markdown is ~~objectively~~ subjectively better than RST, and if you really want to take advantage of some of the fancier Mkdocs features.

The double-edged sword with Mkdocs Material is that it is extremely customisable and unless you're weirdly into documentation and willing to spend ages making a configuration you like you're probably better off sticking with the defaults from Sphinx RTD.

For example [Beanie API](https://beanie-odm.dev/api-documentation/document/), [Segment Geospatial](https://samgeo.gishub.org/common/), [Watchfiles](https://watchfiles.helpmanual.io/api/watch/), etc...

Having said that, note that there is a [**Read the Docs theme**](https://www.mkdocs.org/user-guide/choosing-your-theme/#readthedocs) for mkdocs as well! So if you want to stick to the same style as Sphinx but with markdown that is also an option. From what I've seen it looks basically the same as RTD built with Sphinx, but I have not used the RTD theme so I'm not sure.

**Comment by JamesWrigley** [comment]
I think I tried out the RTD theme but it didn't improve the things that I wanted, mainly the indentation of the members (under the signatures) and the styling of the signatures (bold and with a different color background). But the styling that you've got now is fine I think. @RobertRosca could you push your changes to the `mkdocs` branch? I might play around with it a bit.

> I still have a strong dislike for RST, but if you lean the other way, I don't feel like it's my place to decide that with all the preparations you already did. FWIW it'll look in line with EXtra-data and -geom shrug

Well, I'm also not wedded to RST :) I'm a bit hesitant because it feels less mature than Sphinx, but if the attribute-assignment-statement thingy can be removed then I have no objection. I'll give that another shot.

File: docs/index.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
+EXtra
+=====
+
+EXtra is a library developed at the European XFEL to support *high-level* data
+analysis.
+
+This is where you can find classes and functions for things like:
+
+* Specific analysis techniques (e.g. azimuthal integration or TOF spectrometry)
```
[the remark this answers is not in the document]
Comment (JamesWrigley) [reply docs/index.rst:9]: What do you think about removing those examples entirely? In hindsight maybe we shouldn't mention things that don't exist yet :eyes:

**Comment by JamesWrigley** [comment]
Et voila: https://extra--1.org.readthedocs.build/en/1/index.html

Now I just need to try to remove those ugly CI file paths from the source code viewer...

File: docs/index.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (JamesWrigley) [reply docs/index.rst]: Removed in 981c005.

**Comment by RobertRosca** [comment]
Woah that looks great! :+1:

**Comment by philsmt** [comment]
That looks great, thanks :heart_eyes: 

One question though, do you also see those squares?
image

**Comment by JamesWrigley** [comment]
Hmm, no I don't :thinking: Those are the links to external docs, for me they show up as:
image

Which browser are you using?

**Comment by JamesWrigley** [comment]
(minor fixes in 712c977)

**Comment by philsmt** [comment]
Chrome 114, and it indeed does not happen on Firefox. It's also there on my mobile (with Chrome), so it's not simply a system font issue. Looking into the code it seems to be generated via CSS :flushed:

**Comment by JamesWrigley** [comment]
Derp, it was actually from one of my changes: e630a57
I copied that from somewhere on the mkdocs site and it was using `mask-image`, which is currently unsupported by Chrome. Replaced it with `-webkit-mask-image`, which seems to work in both Firefox and Chrome.

Could you try again? Now it works for me.

**Comment by JamesWrigley** [comment]
Minor changes:
- Set the favicon in 2e92ff4
- Enable permalinks for sections in bd5aed3
- Add `black` in 9fe9790 for formatting the function signatures

File: docs/utils.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+# Utils
```
Comment (philsmt) [inline docs/utils.md:1]: I would either not abbreviate it, use the package name, or both: Utilities `<extra.utils>`

Looking forward, we probably want to include both the human-readable text as well as the package name here. But if you think both is too much, I would probably go with the non-abbreviated human-readable name.

File: docs/index.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Introduction
+EXtra is a library developed at the European XFEL to support *high-level* data
+analysis.
+
+This is where you can find classes and functions for things like:
+
+- Specific analysis techniques
+- High-level components that abstract low-level Karabo devices
+- Random useful things
+
```
[the remark this answers is not in the document]
Comment (JamesWrigley) [reply docs/index.md:12]: Yeah I missed that. I added pages for extra-data and extra-geom in 9786f2b, and reworded the first paragraph to be more one-package-to-rule-them-all-y.

File: docs/index.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+Compared to its sister packages [EXtra-data](https://extra-data.readthedocs.io)
+and [EXtra-geom](https://extra-geom.readthedocs.io), EXtra has a fairly relaxed
```
[the remark this answers is not in the document]
Comment (JamesWrigley) [reply docs/index.md]: I decided to remove the whole paragraph in 9786f2b :stuck_out_tongue: In hindsight I think I wrote it more for developers than for users, and users probably don't need to care about how we develop extra.

File: docs/utils.md (hunk 2 above)
Comment (JamesWrigley) [reply docs/utils.md]: Renamed to `Utilities` in 9786f2b.

File: docs/index.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (JamesWrigley) [reply docs/index.md:12]: Forgot the karabo-bridge documentation, added that in 15cdf11.

**Comment by tmichela** [comment]
Should we use the EuXFEL corporate colors? (cf. [user-docs](https://git.xfel.eu/dataAnalysis/user-documentation/-/blob/main/docs/stylesheets/extra.css))

**Comment by JamesWrigley** [comment]
I think those colors are kinda ugly, so I would prefer not to :see_no_evil:

**Pull request merged**